File: 3-longest-substring-without-repeating-characters/3-longest-substring-without-repeating-characters.cpp
class Solution {
public:
    int lengthOfLongestSubstring(string s) {
        int res=0;
        int i=0,j=0;
        unordered_set<char> set;
        int n=s.size();
        while(i<n && j<n){
            if(set.find(s[j])==set.end()){
                set.insert(s[j++]);
                res=max(res,j-i);
            }else{
                set.erase(s[i++]);
                
            }
        }
        return res;
    }
};

File: 3-longest-substring-without-repeating-characters/3-longest-substring-without-repeating-characters.java
class Solution {
    public int lengthOfLongestSubstring(String s) {
        int i=0,j=0;
        int n= s.length();
        int max=0;
        Set<Character> bucket = new HashSet<>();
        while(i<n && j<n ){
            if(!bucket.contains(s.charAt(j))){
                bucket.add(s.charAt(j++));
                max = Math.max(max,bucket.size());
            }else{
                bucket.remove(s.charAt(i++));
            }
        }
        return max;
    }
}